File: C++/Hard/CompileError/maximum-score-words-formed-by-letters_6.cpp
class Solution {
public:
    int maxScoreWords(vector<string>& words, vector<char>& letters, vector<int>& score) {
        
        unordered_map<char, int> char_map;
        for (auto ch : letters) {
            char_map[ch]++;
        }
        return dfs(words, char_map, score, 0);
    }
    
    int dfs(vector<string>& words, unordered_map<char, int>& char_map, vector<int>& score, int idx) {
        if (idx == words.size()) {
            return 0;
        }
        int res = 0;
        int cur = dfs(words, char_map, score, idx + 1);
        if (isValid(words[idx], char_map)) {
            for (auto ch : words[idx]) {
                char_map[ch]--;
            }
            int score = getScore(words[idx], score);
            res = max(res, score + dfs(words, char_map, score, idx + 1));
            for (auto ch : words[idx]) {
                char_map[ch]++;
            }
        }
        return max(res, cur);
    }
    
    bool isValid(string word, unordered_map<char, int>& char_map) {
        for (auto ch : word) {
            if (char_map[ch] == 0) {
                return false;
            }
        }
        return true;
    }
    
    int getScore(string word, vector<int>& score) {
        int res = 0;
        for (auto ch : word) {
            res += score[ch - 'a'];
        }
        return res;
    }
};

File: Java/Hard/Accepted/maximum-score-words-formed-by-letters_6.java
class Solution {
    public int maxScoreWords(String[] words, char[] letters, int[] score) {
        
        int[] letterCount = new int[26];
        for (char c: letters) {
            letterCount[c - 'a']++;
        }
        return backtrack(words, letterCount, score, 0);
    }
    
    private int backtrack(String[] words, int[] letterCount, int[] score, int start) {
        int maxScore = 0;
        for (int i = start; i < words.length; i++) {
            int[] wordCount = new int[26];
            int wordScore = 0;
            for (char c: words[i].toCharArray()) {
                wordCount[c - 'a']++;
                if (wordCount[c - 'a'] > letterCount[c - 'a']) {
                    wordScore = 0;
                    break;
                }
                wordScore += score[c - 'a'];
            }
            if (wordScore > 0) {
                for (int j = 0; j < 26; j++) {
                    letterCount[j] -= wordCount[j];
                }
                maxScore = Math.max(maxScore, wordScore + backtrack(words, letterCount, score, i + 1));
                for (int j = 0; j < 26; j++) {
                    letterCount[j] += wordCount[j];
                }
            }
        }
        return maxScore;
    }
}